### PR TITLE
Adding bfiocpp recipe

### PR DESCRIPTION
**Pull request by sameeul**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by mkitti** [comment]
This recipe appears to vendor in Tensorstore. That should be packaged separately.

The vendoring is done via `FetchContent_Declare`in CMake:
https://github.com/PolusAI/bfiocpp/blob/ebf37f3211b196cdb81753ffbfe6ff00641cc2bf/CMakeLists.txt#L32-L36

One solution would be to use grayskull to incorporate this recipe from PyPI:
https://pypi.org/project/tensorstore/

**Comment by sameeul** [comment]
The issue is that we are not using the Python package, rather the C++ API of tensorstore. I don't know if that is packaged anywhere. So, not sure what to do other than bringing in the source and adding the license. 

Any other ideas?

**Comment by mkitti** [comment]
Basically, you need to package Tensorstore as its own recipe first. Then have this package depend on that one.

The complicated thing is that Tensorstore itself also vendors many libraries in, but hopefully many of those are packaged already.

**Comment by sameeul** [comment]
This is what their documentation says about C++ integration 

https://google.github.io/tensorstore/installation.html#cmake-integration

I am trying to understand what is the issue with bringing in the source and compiling the targets as part of this build process. Can you please explain the issue?

**Comment by mkitti** [comment]
Within conda-forge the packages are supposed to be modular where code is not duplicated. The idea is to maintain compatibility between packages with common dependencies as well as their corresponding configurations.

There is a tensorstore recipe for conda-forge:

https://github.com/conda-forge/tensorstore-feedstock/blob/main/recipe%2Fbuild.sh

I see that tensorstore is likely not meant to be used as a shared library itself, but you should consider taking inspiration from the above build script to use as many "system" libraries as possible.

If this is not something you want to do, you could always consider uploading your package as is to your own channel.

**Comment by sameeul** [comment]
Thanks for pointing that out. I was also experimenting with that locally. From the auto-generated CMake files, I see the following: 
```
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_ABSL:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_AOM:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_AVIF:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_BENCHMARK:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_BLAKE3:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_BLOSC:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_BROTLI:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_BZIP2:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_CARES:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_CURL:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_DAV1D:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_GTEST:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_HALF:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_JPEG:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_LIBLZMA:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_LIBYUV:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_LZ4:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_NGHTTP2:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_NLOHMANN_JSON:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_OPENSSL:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_PNG:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_PROTOBUF:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_RE2:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_RIEGELI:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_SNAPPY:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_TIFF:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_WEBP:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_ZLIB:BOOL=OFF
build_man/CMakeCache.txt:TENSORSTORE_USE_SYSTEM_ZSTD:BOOL=OFF
```
I will try to use as many of them possible from the existing packages. I will push an update soon.

**Comment by sameeul** [comment]
@mkitti : I have taken your suggestion and updated the recipe. I tried to follow the `tensorstore` recipe and use as many as system libraries possible. Can you please take a look?

**Comment by mkitti** [comment]
Thank you. This looks better to me.

@conda-forge/help-c-cpp, could you take a look?

File: recipes/bfiocpp/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    - libprotobuf
+    - benchmark
+    - pybind11
+    # this is header-only, so should be OK
+    - nlohmann_json 3.10.5
+    - libavif
+
+  run:
+    - python
+    - {{ pin_compatible('numpy') }}
```
Comment (xhochy) [inline recipes/bfiocpp/meta.yaml:55]: `numpy` has a `run_exports` and thus this should no longer be used.
```suggestion
```

File: recipes/bfiocpp/bld.bat [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+set CMAKE_USE_SYSTEM_LIBS=%CMAKE_USE_SYSTEM_LIBS% -DTENSORSTORE_USE_SYSTEM_SNAPPY=ON
+set CMAKE_USE_SYSTEM_LIBS=%CMAKE_USE_SYSTEM_LIBS% -DTENSORSTORE_USE_SYSTEM_TIFF=ON
+set CMAKE_USE_SYSTEM_LIBS=%CMAKE_USE_SYSTEM_LIBS% -DTENSORSTORE_USE_SYSTEM_WEBP=ON
+set CMAKE_USE_SYSTEM_LIBS=%CMAKE_USE_SYSTEM_LIBS% -DTENSORSTORE_USE_SYSTEM_ZLIB=ON
+set CMAKE_USE_SYSTEM_LIBS=%CMAKE_USE_SYSTEM_LIBS% -DTENSORSTORE_USE_SYSTEM_ZSTD=ON
+set CMAKE_USE_SYSTEM_LIBS=%CMAKE_USE_SYSTEM_LIBS% -DAVIF_ENABLE_GTEST=OFF
+set CMAKE_USE_SYSTEM_LIBS=%CMAKE_USE_SYSTEM_LIBS% -DBENCHMARK_ENABLE_GTEST_TESTS=OFF
+set CMAKE_USE_SYSTEM_LIBS=%CMAKE_USE_SYSTEM_LIBS% -DINSTALL_GTEST=OFF
+
+set CMAKE_ARGS=-DCMAKE_PREFIX_PATH=%PREFIX% -DCMAKE_INSTALL_PREFIX=%PREFIX% %CMAKE_USE_SYSTEM_LIBS%
```
Comment (xhochy) [inline recipes/bfiocpp/bld.bat:31]: conda-forge already comes with a pre-set CMAKE_ARGS that contains the necessary values
```suggestion
set CMAKE_ARGS=%CMAKE_ARGS% %CMAKE_USE_SYSTEM_LIBS%
```

File: recipes/bfiocpp/build.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+CMAKE_USE_SYSTEM_LIBS+=" -DTENSORSTORE_USE_SYSTEM_SNAPPY=ON" 
+CMAKE_USE_SYSTEM_LIBS+=" -DTENSORSTORE_USE_SYSTEM_TIFF=ON" 
+CMAKE_USE_SYSTEM_LIBS+=" -DTENSORSTORE_USE_SYSTEM_WEBP=ON" 
+CMAKE_USE_SYSTEM_LIBS+=" -DTENSORSTORE_USE_SYSTEM_ZLIB=ON" 
+CMAKE_USE_SYSTEM_LIBS+=" -DTENSORSTORE_USE_SYSTEM_ZSTD=ON" 
+CMAKE_USE_SYSTEM_LIBS+=" -DAVIF_ENABLE_GTEST=OFF" 
+CMAKE_USE_SYSTEM_LIBS+=" -DBENCHMARK_ENABLE_GTEST_TESTS=OFF" 
+CMAKE_USE_SYSTEM_LIBS+=" -DINSTALL_GTEST=OFF" 
+
+CMAKE_ARGS="-DCMAKE_PREFIX_PATH=$PREFIX -DCMAKE_INSTALL_PREFIX=$PREFIX $CMAKE_USE_SYSTEM_LIBS" python -m pip install . -vvv
```
Comment (xhochy) [inline recipes/bfiocpp/build.sh:41]: conda-forge already comes with a pre-set CMAKE_ARGS that contains the necessary values
```suggestion
CMAKE_ARGS="$CMAKE_ARGS $CMAKE_USE_SYSTEM_LIBS" python -m pip install . -vvv
```

**Review by xhochy — APPROVED** [review]

**Pull request merged**